core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -534,9 +538,30 @@ class TransactionStateManager(brokerId: Int,
             // return NOT_COORDINATOR to let the client re-discover the transaction coordinator
             info(s"Updating $transactionalId's transaction state (txn topic partition ${partitionFor(transactionalId)}) to $newMetadata with coordinator epoch $coordinatorEpoch for $transactionalId " +
               s"failed after the transaction message has been appended to the log since the corresponding metadata does not exist in the cache anymore")
-
             responseError = Errors.NOT_COORDINATOR
         }
+      } else {
+        // Reset the pending state when returning an error, since there is no active transaction for the transactional id at this point.
+        getAndMaybeAddTransactionState(transactionalId) match {
+          case Right(Some(epochAndTxnMetadata)) =>
+            val metadata = epochAndTxnMetadata.transactionMetadata
+            metadata synchronized {
+              if (epochAndTxnMetadata.coordinatorEpoch == coordinatorEpoch) {
+                debug(s"TransactionalId ${metadata.transactionalId}, resetting pending state since we are returning error $responseError")
+                metadata.pendingState = None
+              } else {
+                info(s"TransactionalId ${metadata.transactionalId} coordinator epoch changed from " +
+                  s"${epochAndTxnMetadata.coordinatorEpoch} to $coordinatorEpoch after append to log returned $responseError")
+              }
+            }
+          case Right(None) =>
+            // Do nothing here, since we want to return the original append error to the user.
+            warn(s"Found no metadata TransactionalId $transactionalId after append to log returned error $responseError")
```

Comment:
I think both of these two cases `Right(None)` and `Left(error)` are actually expected scenario in practice, so we should not log it as `warn`?